plugin.js

```diff
@@ -45,60 +45,274 @@
    * @param {string} url The video URL
    * @returns {Video} The video data
    */
-  function parseUrl (url) {
-    var data = {}
-    data.url = url
+  function parseUrl(url) {
+    const data = {};
 
-    PROVIDERS.forEach(function (name) {
-      var provider = PROVIDERS[name]
-      var matches = url.match(provider.pattern)
+    PROVIDERS.forEach(function buildData(name) {
+      const provider = PROVIDERS[name];
+      const matches = url.match(provider.pattern);
       if (matches) {
-        data.provider = name
-        data.id = matches[1]
-        data.url = provider.api + data.id
-        var queries = matches[2] || ''
-        queries.split('&').forEach(function (querystring) {
-          var kv = querystring.split('=')
-          data[kv[0]] = kv[1]
-        })
+        data.name = name;
+        const id = matches[1];
+        const providerUrl = provider.api + id;
+        data.url = providerUrl;
+        return data;
       }
-    })
+    });
 
-    return data
+    //if data is an empty object, then it is not a known provider
+    if (Object.keys(data).length === 0) {
+      return false;
+    } else {
+      return data;
+    }
   }
 
-  function Plugin (editor, url) {
-    this.showDialog = this.showDialog.bind(this)
-    this.render = this.render.bind(this)
+  function Plugin(editor, url) {
+    this.showDialog = this.showDialog.bind(this);
+    this.render = this.render.bind(this);
+    const plugin = this;
+    plugin.editor = editor;
+    plugin.currentProvider = null;
+    this.configs = {};
+
+    this.configs["unknown"] = {
+      title: "Insert a video",
```

Comment:
There are a lot of duplication here. We may want to have a single object for config, and for each providers, use spread operator to extend/override settings that are specific to them.